<a href="https://colab.research.google.com/github/aditya002ai/MachineLearning/blob/main/multicollinearity_in_linear_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [18]:
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

In [11]:
df_adv = pd.read_csv('/content/Advertising.csv',index_col=0)


In [9]:
X = df_adv[['TV','radio','newspaper']]
Y= df_adv['sales']

In [10]:
df_adv.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

y = b0 + b1(TV) +b2(radio) +b3(newspaper)
OLS =  Ordinary list square

In [12]:
X = sm.add_constant(X)

In [14]:
X

const     TV  radio  newspaper
1      1.0  230.1   37.8       69.2
2      1.0   44.5   39.3       45.1
3      1.0   17.2   45.9       69.3
4      1.0  151.5   41.3       58.5
5      1.0  180.8   10.8       58.4
..     ...    ...    ...        ...
196    1.0   38.2    3.7       13.8
197    1.0   94.2    4.9        8.1
198    1.0  177.0    9.3        6.4
199    1.0  283.6   42.0       66.2
200    1.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [15]:
model = sm.OLS(Y,X).fit()

In [17]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sun, 27 Jun 2021   Prob (F-statistic):           1.58e-96
Time:                        14:33:57   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Now const is b0, TV is b1 , radio is b2, newspaper is b3  , 
(newspaper -0.0010	0.006) negative value means you do not have to do that much expanditure

****R-squared  (0-1)*** if it is near to 1 means predict good model
***std err***  if this number is less then no multicollinearity

***Note*** This model is not having multicollinearity because R-sqared value is near to 1 adn std err value is very less.

In [19]:
X.iloc[:,1:].corr()

TV     radio  newspaper
TV         1.000000  0.054809   0.056648
radio      0.054809  1.000000   0.354104
newspaper  0.056648  0.354104   1.000000

In [20]:
df_salary = pd.read_csv('/content/Salary_Data.csv')
df_salary.head()

YearsExperience   Age  Salary
0              1.1  21.0   39343
1              1.3  21.5   46205
2              1.5  21.7   37731
3              2.0  22.0   43525
4              2.2  22.2   39891

In [21]:
X = df_salary[['YearsExperience','Age']]
Y =df_salary['Salary']

In [22]:
X = sm.add_constant(X)

In [23]:
model = sm.OLS(Y,X).fit()

In [24]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Sun, 27 Jun 2021   Prob (F-statistic):           1.35e-19
Time:                        14:56:52   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Here p value of Age is > than .05 and st error is also value heigh, so it have multicollinearity.

In [25]:
X.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

for this you can see that there is 98% correlation between Age and Years of Experience, so which one you can drop is, so simple way you can see the p values of Age is greater so you can drop Age.